cdap-master/src/main/java/co/cask/cdap/data/tools/flow/FlowQueuePendingCorrector.java

```diff
@@ -0,0 +1,229 @@
+import com.google.inject.Injector;
+import com.google.inject.Singleton;
+import org.apache.commons.cli.BasicParser;
+import org.apache.commons.cli.CommandLine;
+import org.apache.commons.cli.CommandLineParser;
+import org.apache.commons.cli.HelpFormatter;
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.Options;
+import org.apache.commons.cli.ParseException;
+import org.apache.twill.kafka.client.KafkaClientService;
+import org.apache.twill.zookeeper.ZKClientService;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Corrects the "queue.pending" metric emitted in {@link co.cask.cdap.internal.app.runtime.flow.FlowletProgramRunner}.
+ */
+public class FlowQueuePendingCorrector extends AbstractIdleService {
+
+  private static final Gson GSON = new Gson();
+
+  private final MetricsCollectionService metricsCollectionService;
+  private final MetricStore metricStore;
+  private final KafkaClientService kafkaClientService;
+  private final HBaseQueueDebugger queueDebugger;
+  private final ZKClientService zkClientService;
+
+  @Inject
+  public FlowQueuePendingCorrector(HBaseQueueDebugger queueDebugger, ZKClientService zkClientService,
+                                   MetricsCollectionService metricsCollectionService, MetricStore metricStore,
+                                   KafkaClientService kafkaClientService) {
+    this.queueDebugger = queueDebugger;
+    this.zkClientService = zkClientService;
+    this.metricsCollectionService = metricsCollectionService;
+    this.metricStore = metricStore;
+    this.kafkaClientService = kafkaClientService;
+  }
+
+  public void run(Id.Flow flow, String producerFlowlet, String consumerFlowlet,
+                  String flowletQueue) throws Exception {
+
+    QueueName queueName = QueueName.fromFlowlet(flow, producerFlowlet, flowletQueue);
+    long consumerGroupId = FlowUtils.generateConsumerGroupId(flow, consumerFlowlet);
+    HBaseQueueDebugger.QueueStatistics stats = queueDebugger.scanQueue(queueName, consumerGroupId);
+    long correctQueuePendingValue = stats.getUnprocessed() + stats.getProcessedAndNotVisible();
+
+    Map<String, String> tags = ImmutableMap.<String, String>builder()
+      .put(Constants.Metrics.Tag.NAMESPACE, flow.getNamespaceId())
+      .put(Constants.Metrics.Tag.APP, flow.getApplicationId())
+      .put(Constants.Metrics.Tag.FLOW, flow.getId())
+      .put(Constants.Metrics.Tag.CONSUMER, consumerFlowlet)
+      .put(Constants.Metrics.Tag.PRODUCER, producerFlowlet)
+      .put(Constants.Metrics.Tag.FLOWLET_QUEUE, flowletQueue)
+      .build();
+
+    MetricDataQuery query = new MetricDataQuery(
+      0, 0, Integer.MAX_VALUE, 1, ImmutableMap.of("system.queue.pending", AggregationFunction.SUM),
+      tags, ImmutableList.<String>of(), null);
+
+    Collection<MetricTimeSeries> results = metricStore.query(query);
+    System.out.println("Got results: " + GSON.toJson(results));
+
+    if (results.isEmpty()) {
```

Comment:
But you should make sure that the flow, consumer and producer flowlet and queue actually exist. Otherwise you would emit a metric for a queue that does not exist.